File: python/Two Pointers/two_sum_two.py
from typing import List

class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        """
        Find two numbers in a sorted array that add up to the target using two pointers.
        Returns 1-based indices.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # Initialize two pointers: left at start, right at end
        l, r = 0, len(numbers) - 1
        
        # Continue until pointers meet
        while l < r:
            # If sum is greater than target, move right pointer left
            # (to get a smaller sum)
            if numbers[l] + numbers[r] > target:
                r = r - 1
            # If sum is less than target, move left pointer right
            # (to get a larger sum)
            elif numbers[l] + numbers[r] < target:
                l = l + 1
            # If sum equals target, return 1-based indices
            else:
                return [l+1, r+1]
        raise ValueError("No solution found")

# -------------------
# Test Cases
# -------------------
def test_two_sum():
    s = Solution()
    # Test case 1: Example
    numbers = [2,7,11,15]
    target = 9
    assert s.twoSum(numbers, target) == [1,2]
    # Test case 2: Negative numbers
    numbers = [-3,0,3,4]
    target = 0
    assert s.twoSum(numbers, target) == [1,3]
    # Test case 3: Duplicates
    numbers = [1,1,2,3]
    target = 2
    assert s.twoSum(numbers, target) == [1,2]
    # Test case 4: Large input
    numbers = [1,2,3,4,4,9,56,90]
    target = 8
    assert s.twoSum(numbers, target) == [4,5]
    # Test case 5: Only two elements
    numbers = [1,2]
    target = 3
    assert s.twoSum(numbers, target) == [1,2]
    print("All test cases passed!")

if __name__ == "__main__":
    test_two_sum()
            

File: C++/Two Pointers/two_sum_two.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Find two numbers in sorted array that sum to target
     * Return 1-based indices of the two numbers
     * 
     * Approach:
     * - Use two pointers: left at start, right at end
     * - If sum > target, move right pointer left (decrease sum)
     * - If sum < target, move left pointer right (increase sum)
     * - If sum = target, return indices (1-based)
     * 
     * Time: O(n) - single pass through array
     * Space: O(1) - only two pointers used
     */
    vector<int> twoSum(vector<int>& numbers, int target) {
        // Initialize pointers
        int left = 0;
        int right = numbers.size() - 1;
        
        while (left < right) {
            int currSum = numbers[left] + numbers[right];
            
            if (currSum > target) {
                right--;  // Need smaller sum
            }
            else if (currSum < target) {
                left++;   // Need larger sum
            }
            else {
                // Found solution, return 1-based indices
                return {left + 1, right + 1};
            }
        }
        
        // No solution found (though problem guarantees one exists)
        return {};
    }
};

/*
 * Detailed C++ Syntax Explanation:
 * 
 * 1. Vector Operations:
 *    numbers.size()   // Get array length
 *    numbers[left]    // Access element at index
 * 
 * 2. Pointer/Index Variables:
 *    int left = 0    // Start of array
 *    int right = numbers.size() - 1  // End of array
 * 
 * 3. While Loop:
 *    while (left < right)  // Continue until pointers meet
 * 
 * 4. Return Statement:
 *    return {left + 1, right + 1}  // Returns vector from initializer list
 *    return {};  // Returns empty vector
 * 
 * 5. Vector<int>:
 *    - Return type specifying vector of integers
 *    - Used for returning pair of indices
 */
